pkg/apps/alert/alert.go

```diff
@@ -117,33 +118,133 @@ func (a Alert) Ensure(alt *alertapi.Alert) error {
+	var namespace string
+	if len(values) == 0 {
+		return fmt.Errorf("unable to find the Black Duck namespace")
+	} else if len(values) == 1 {
+		name = values[0]
+		namespace = values[0]
+	} else {
+		name = values[1]
+		namespace = values[0]
+	}
+
 	var err error
 	// Verify whether the namespace exist
 	_, err = util.GetNamespace(a.kubeClient, namespace)
 	if err != nil {
 		return fmt.Errorf("unable to find the namespace %+v due to %+v", namespace, err)
 	}
-	// Delete the namespace
-	err = util.DeleteNamespace(a.kubeClient, namespace)
+
+	// get all replication controller for the namespace
+	rcs, err := util.ListReplicationControllers(a.kubeClient, namespace, fmt.Sprintf("app!=alert,name!=%s", name))
+	if err != nil {
+		return errors.Annotatef(err, "unable to list the replication controller %+v", namespace)
+	}
+
+	// get all deployment for the namespace
+	deployments, err := util.ListDeployments(a.kubeClient, namespace, fmt.Sprintf("app!=alert,name!=%s", name))
 	if err != nil {
-		return fmt.Errorf("unable to delete the namespace %+v due to %+v", namespace, err)
+		return errors.Annotatef(err, "unable to list the deployments in %s namespace", namespace)
 	}
-	// Verify whether the namespace deleted
-	var attempts = 30
-	var retryWait time.Duration = 10
-	for i := 0; i <= attempts; i++ {
-		_, err := util.GetNamespace(a.kubeClient, namespace)
+
+	// get only Alert related replication controller for the namespace
+	alertRCs, err := util.ListReplicationControllers(a.kubeClient, namespace, fmt.Sprintf("app=alert,name=%s", name))
+	if err != nil {
+		return errors.Annotatef(err, "unable to list the Alert's replication controller in %s", namespace)
+	}
+
+	// if both the length same, then delete the namespace, if different, delete only the replication controller
+	if (len(rcs.Items) == 0 && len(deployments.Items) == 0) || (len(rcs.Items) == len(alertRCs.Items)) {
+		// Delete the namespace
+		err = util.DeleteNamespace(a.kubeClient, namespace)
+		if err != nil {
+			return errors.Annotatef(err, "unable to delete namespace %s", namespace)
+		}
+
+		// Verify whether the namespace deleted
+		var attempts = 30
+		var retryWait time.Duration = 10
+		for i := 0; i <= attempts; i++ {
+			_, err := util.GetNamespace(a.kubeClient, namespace)
+			if err != nil {
+				log.Infof("Deleted the namespace %+v", namespace)
+				break
+			}
+			if i >= 10 {
+				return fmt.Errorf("unable to delete the namespace %+v after %f minutes", namespace, float64(attempts)*retryWait.Seconds()/60)
+			}
+			time.Sleep(retryWait * time.Second)
+		}
+	} else {
+		// delete the replication controller
+		for _, alertRC := range alertRCs.Items {
+			err = util.DeleteReplicationController(a.kubeClient, namespace, alertRC.GetName())
+			if err != nil {
+				return errors.Annotatef(err, "unable to delete the %s replication controller in %s namespace", alertRC.GetName(), namespace)
+			}
+		}
+
+		// get only Black Duck related services for the namespace
```

Comment:
Probably meant `Alert` in all the cases in this file.

Reply:
fixed